// @eui/lit-component/index.js
/**
 * @license
 * COPYRIGHT Ericsson 2022
 * The copyright to the computer program(s) herein is the property of
 * Ericsson Inc. The programs may be used and/or copied only with written
 * permission from Ericsson Inc. or in accordance with the terms and
 * conditions stipulated in the agreement/contract under which the
 * program(s) have been supplied.
 */
// aggregating the exports from component/lit-component/lit-html and lit-html's directives
// in order to make them available from lit-component

export { LitComponent, LitStateless } from './src/lit-component.js';
export { html, render, svg } from 'lit-html';

// Note: The following lit-html directives are a potential source of XSS
// as per their docs if not used correctly. Do not export them.
// E-UI SDK will not be a vector for UI attacks
// - unsafeHTML
// - unsafeSVG
// - templateContent

// export lit-html directives
export { asyncAppend } from 'lit-html/directives/async-append.js';
export { asyncReplace } from 'lit-html/directives/async-replace.js';
export { cache } from 'lit-html/directives/cache.js';
export { classMap } from 'lit-html/directives/class-map.js';
export { choose } from 'lit-html/directives/choose.js';
export { createRef } from 'lit-html/directives/ref.js';
export { guard } from 'lit-html/directives/guard.js';
export { ifDefined } from 'lit-html/directives/if-defined.js';
export { join } from 'lit-html/directives/join.js';
export { keyed } from 'lit-html/directives/keyed.js';
export { live } from 'lit-html/directives/live.js';
export { map } from 'lit-html/directives/map.js';
export { range } from 'lit-html/directives/range.js';
export { ref } from 'lit-html/directives/ref.js';
export { repeat } from 'lit-html/directives/repeat.js';
export { styleMap } from 'lit-html/directives/style-map.js';
export { until } from 'lit-html/directives/until.js';
export { when } from 'lit-html/directives/when.js';
export { nothing } from 'lit-html';

export { definition } from '@eui/component';
