// @eui/base/dist/button.js
import { definition, LitComponent, html, classMap, ifDefined } from '@eui/lit-component';
import { Icon } from '@eui/theme/icon';

var style = ":host {\n  display:inline-block;\n  text-rendering: optimizeLegibility;\n  -webkit-font-smoothing: antialiased;\n}\n:host([fullwidth]){\n  display: block;\n}\n\nbutton,\na {\n  display: flex;\n  width: 100%;\n  align-items: center;\n  flex-direction: row;\n  padding: 5px 12px 4px;\n  border-radius: var(--btn-radius, 3px);\n  outline: 0;\n  font-size: var(--btn-font-size, 14px);\n  color: var(--btn-secondary-text, #242424);\n  background-color: var(--btn-secondary-background, transparent);\n  border: 1px solid var(--btn-secondary-border, #242424);\n  line-height: var(--btn-line-height, 21px);\n  box-sizing: border-box;\n  background-image: var(--btn-secondary-gradient);\n  font-family: var(--font-main, \"Ericsson Hilda\", \"Helvetica\", \"sans-serif\");\n}\n\na {\n  text-decoration: none;\n}\n\n.button__icon {\n  display: inline-block;\n}\n\n.with--icon {\n  margin-left: var(--space-base, 8px);\n  margin-right: 0;\n  text-align: left;\n  min-height: var(--btn-line-height, 21px);\n}\n\n.button__label {\n  flex: auto;\n  overflow: hidden;\n  white-space: nowrap;\n  text-overflow: ellipsis;\n}\n\n.btn--align--edge {\n  display: inline-flex;\n}\n\n.btn--align--edge .button__label.with--icon {\n  text-align: right;\n}\n\n.btn--align--edge .button__label {\n  text-align: left;\n}\n\n.btn--reverse .button__label {\n  text-align: right;\n}\n\n.btn--reverse.btn--align--edge .button__label {\n  text-align: right;\n}\n\n.btn--reverse.btn--align--edge .button__label.with--icon {\n  text-align: left;\n}\n\n.btn--reverse .with--icon {\n  margin-left: 0;\n  margin-right: var(--space-base, 8px);\n  flex: auto;\n}\n\n.btn--reverse eui-icon {\n  text-align: left;\n}\n\n.btn--align--edge eui-icon {\n  text-align: left;\n  flex: none;\n}\n\n.btn--reverse.btn--align--edge eui-icon {\n  text-align: right;\n  flex: none;\n}\n\neui-icon {\n  flex: auto;\n  text-align: right;\n  --icon-color: var(--btn-secondary-text, #242424);\n}\n\n:host(:empty) .with--icon{\n  margin-left: 0;\n}\n\n.btn:last-child {\n  margin-right: 0;\n}\n\n:host([disabled]) a,\n:host([disabled]) button{\n  opacity: 0.4;\n  cursor: default;\n}\n\n:host(:hover:not([disabled])) .btn {\n  color: var(--btn-secondary-hover-text, #FFF);\n  text-decoration: none;\n  background-color: var(--btn-secondary-hover-background, #4E4E4E);\n  border: 1px solid var(--btn-secondary-hover-border, #4E4E4E);\n  background-image: var(--btn-secondary-hover-gradient);\n  --icon-color: var(--btn-secondary-hover-text, #FFF);\n  cursor: pointer;\n}\n:host(:hover:not([disabled])) eui-icon {\n  --icon-color: var(--btn-secondary-hover-text, #FFF);\n}\n\n:host(:not([disabled])) .btn:active {\n  color: var(--btn-secondary-active-text, #FFF);\n  background-color: var(--btn-secondary-active-background, #242424);\n  border: 1px solid var(--btn-secondary-active-border, #242424);\n  box-shadow: var(--btn-secondary-active-box-shadow);\n  background-image: var(--btn-secondary-active-gradient);\n  -webkit-transition: none;\n  transition: none;\n}\n\n:host(:not([disabled])) .btn:active eui-icon{\n  --icon-color: var(--btn-secondary-active-text, #FFF);\n}\n\n:host([primary]) eui-icon {\n  --icon-color: var(--btn-primary-text, #FFF);\n}\n\n:host([warning]) eui-icon {\n  --icon-color: var(--btn-warning-text, #FFF);\n}\n\n:host([primary]) .btn {\n  color: var(--btn-primary-text, #FFF);\n  background-color: var(--btn-primary-background, #1174E6);\n  border: 1px solid var(--btn-primary-border, #1174E6);\n  background-image: var(--btn-color-gradient);\n}\n\n:host(:hover:not([disabled])[primary]) .btn {\n  color: var(--btn-primary-text, #FFF);\n  background-color: var(--btn-primary-hover-background, #0069C2);\n  border: 1px solid var(--btn-primary-hover-border, #0069C2);\n  background-image: var(--btn-color-hover-gradient);\n}\n:host(:hover:not([disabled])[primary]) eui-icon{\n  --icon-color: var(--btn-primary-text, #FFF);\n}\n\n:host(:not([disabled])[primary]) .btn:active {\n  color: var(--btn-primary-text, #FFF);\n  background-color: var(--btn-primary-active-background, #005499);\n  border: 1px solid var(--btn-primary-active-border, #005499);\n  box-shadow: var(--btn-color-active-box-shadow);\n  background-image: var(--btn-color-active-gradient);\n  -webkit-transition: none;\n  transition: none;\n}\n\n:host(:not([disabled])[primary]) .btn:active eui-icon {\n  --icon-color: var(--btn-primary-text, #FFF);\n}\n\n:host([warning]) .btn {\n  color: var(--btn-warning-text, #FFF);\n  background-color: var(--btn-warning-background, #ED0E00);\n  border: 1px solid var(--btn-warning-border, #ED0E00);\n  background-image: var(--btn-color-gradient);\n}\n\n:host(:hover:not([disabled])[warning]) .btn {\n  color: var(--btn-warning-text, #FFF);\n  background-color: var(--btn-warning-hover-background, #D40D02);\n  border: 1px solid var(--btn-warning-hover-border, #D40D02);\n  background-image: var(--btn-color-hover-gradient);\n}\n:host(:hover:not([disabled])[warning]) eui-icon {\n  --icon-color: var(--btn-warning-text, #FFF);\n}\n\n:host(:not([disabled])[warning]) .btn:active {\n  color: var(--btn-warning-text, #FFF);\n  background-color: var(--btn-warning-active-background, #BB0B02);\n  border: 1px solid var(--btn-warning-active-border, #BB0B02);\n  box-shadow: var(--btn-color-active-box-shadow);\n  background-image: var(--btn-color-active-gradient);\n  -webkit-transition: none;\n  transition: none;\n}\n:host(:not([disabled])[warning]) .btn:active eui-icon {\n  color: var(--btn-warning-text, #FFF);\n}\n\nlabel input[type=\"file\"] {\n  display: none;\n}\n\n:host(:focus) {\n  outline: none;\n}\n\n:focus {\n  outline: 1px solid var(--purple, #a56ebe);\n  outline-offset: 1px;\n}\n";

/**
 * @license
 * COPYRIGHT Ericsson 2022
 * The copyright to the computer program(s) herein is the property of
 * Ericsson Inc. The programs may be used and/or copied only with written
 * permission from Ericsson Inc. or in accordance with the terms and
 * conditions stipulated in the agreement/contract under which the
 * program(s) have been supplied.
 */

class Button extends LitComponent {
  static get components() {
    return {
      'eui-icon': Icon,
    };
  }

  constructor() {
    super();
    this.inFocus = false;
    this.addEventListener('click', this);
    this.addEventListener('focus', () => {
      // Shift focus to button field to avoid double tab
      // Clear tabindex to avoid having to press shift+tab to get out. Restored on focus loss.
      this.shadowRoot.querySelector('.btn').focus();
      this.inFocus = true;
      this.tabindex = -1;
    });
  }

  didUpgrade() {
    this.userDefinedTabIndex = this.tabindex;
    if (this.disabled) {
      this.tabindex = -1;
    }
  }

  didChangeProps(changedProps) {
    if (changedProps.has('tabindex') && !this.inFocus && !this.disabled) {
      this.userDefinedTabIndex = this.tabindex;
    }
    if (changedProps.has('disabled')) {
      if (this.disabled) {
        this.tabindex = -1;
      } else {
        this.tabindex = this.userDefinedTabIndex;
      }
    }
  }

  /**
   * If the button is disabled then stop propagation of the event
   * @param {Event} event - event from component.
   */
  handleEvent(event) {
    if (this.disabled) {
      event.preventDefault();
      event.stopImmediatePropagation();
    } else if (event.type === 'blur') {
      this.inFocus = false;
      this.tabindex = this.userDefinedTabIndex;
    }
  }

  get buttonIcon() {
    return this.shadowRoot.querySelector('eui-icon');
  }

  /**
   * _isIcon determines if there is an icon added to the button.
   *
   * @function _isIcon
   * @private
   */
  _isIcon = () => this.icon && this.icon.length > 0;

  /**
   * _withIcon renders the markup for the icon if it is added to the button
   *
   * @function _withIcon
   * @private
   */
  _withIcon = () => {
    if (this._isIcon()) {
      return html`<eui-icon name=${this.icon} size="14px"></eui-icon>`;
    }
    return null;
  };

  /**
   * create classMap
   *
   * @returns Object - classMap
   * @private
   */
  /* prettier-ignore */
  _setClasses = () => classMap({
      btn: true,
      'btn--reverse': this.reverse,
      'btn--align--edge': this.alignEdge,
    });

  /**
   * Render markup for link
   *
   * @function _renderLink
   * @private
   */
  _renderLink = () => html`
    <a
      class="${this._setClasses()}"
      href="${this.href}"
      download=${ifDefined(this.download)}
      target="${ifDefined(this.target)}"
      ?disabled=${this.disabled}
      @blur=${this}
    >
      ${this.reverse ? '' : this._withIcon()}
      <span class="button__label ${this._isIcon() ? 'with--icon' : ''}">
        <slot></slot>
      </span>
      ${this.reverse ? this._withIcon() : ''}
    </a>
  `;

  /**
   * Render markup for the regular button
   *
   * @function _renderButton
   * @private
   */
  _renderButton = () => html`
    <button
      class="${this._setClasses()}"
      ?disabled=${this.disabled}
      @blur=${this}
    >
      ${this.reverse ? '' : this._withIcon()}
      <span class="button__label ${this._isIcon() ? 'with--icon' : ''}">
        <slot></slot>
      </span>
      ${this.reverse ? this._withIcon() : ''}
    </button>
  `;

  /**
   * Render the button component. This function is called each time a
   * prop changes.
   */
  render() {
    if (this.href) {
      return this._renderLink();
    }
    return this._renderButton();
  }
}

definition('eui-button', {
  style,
  props: {
    alignEdge: { attribute: true, type: Boolean },
    disabled: { attribute: true, type: Boolean },
    download: { attribute: true, type: String },
    fullwidth: { attribute: true, type: Boolean },
    href: { attribute: true, type: String },
    icon: { attribute: true, type: String },
    reverse: { attribute: true, type: Boolean },
    primary: { attribute: true, type: Boolean },
    target: { attribute: true, type: String },
    tabindex: { attribute: true, type: Number },
    warning: { attribute: true, type: Boolean },
  },
})(Button);

export { Button };
