// @eui/component/index.js
/**
 * @license
 * COPYRIGHT Ericsson 2022
 * The copyright to the computer program(s) herein is the property of
 * Ericsson Inc. The programs may be used and/or copied only with written
 * permission from Ericsson Inc. or in accordance with the terms and
 * conditions stipulated in the agreement/contract under which the
 * program(s) have been supplied.
 */
import StatelessComponentMixin from './src/Stateless.js';
import StatefulComponentMixin from './src/Component.js';
import { CustomElementsMixin } from './src/mixins/scoped-elements-mixin.js';
import Template from './src/renderers/Template.js';

export class Stateless extends StatelessComponentMixin(HTMLElement) {}
export class Component extends StatefulComponentMixin(HTMLElement) {}
export class TemplateComponent extends CustomElementsMixin(
  Template(StatefulComponentMixin(HTMLElement)),
) {}

export { prop, registerMixin, hasMixin } from './src/utils/exports.js';
export { definition, APPEND, PREPEND } from './src/utils/definition.js';
export { StoreMixin as Store } from './src/mixins/Store.js';
