// @eds/vanilla/notification-log/NotificationLog.js
import { Notification } from '../notification/Notification';

export const NotificationLog = {

  /**
   * Create component.
   * @public
   */
  init() {
    this.dom = {
      notificationLog: document.querySelector('.notification-log'),
      notificationLogContainer: document.querySelector('.notification-log-container'),
      notificationLogTrigger: document.querySelector('.notification-log-trigger'),
      notificationLogNumber: document.querySelector('.notification-log-number'),
      notificationLogEmpty: document.querySelector('.notification-log-empty'),
    };
    this.state = {
      isNew: 0,
      seen: false,
      notifications: [],
      notificationsDOM: [],
      liveNotification: null,
    };
    this.listeners = {
      clickTrigger: () => this._clickTrigger(),
      SyspanelOpenEvent: () => this._removeLiveNotification(),
      SyspanelTabEvent: () => this._clearAllNotifications(),
      SyspanelCloseEvent: () => this._clearAllNotifications(),
    };
    this.hasActiveListeners = [];
    this.events = {
      toggleSyspanel: new CustomEvent('toggleSyspanel', { detail: this.dom.notificationLog }),
      NotificationLogCounterUpdate: new CustomEvent('NotificationLogCounterUpdate'),
    };
    this.notificationTimeout = 10000;
    this._addEventListeners();
  },

  /**
   * Destroy component.
   * @public
   */
  destroy() {
    this._removeEventListeners();
  },

  /**
   * Add event listeners.
   * @private
   */
  _addEventListeners() {
    this.dom.notificationLogTrigger.addEventListener('click', this.listeners.clickTrigger);
    window.addEventListener('SyspanelOpenEvent', this.listeners.SyspanelOpenEvent);
    window.addEventListener('SyspanelTabEvent', this.listeners.SyspanelTabEvent);
    window.addEventListener('SyspanelCloseEvent', this.listeners.SyspanelCloseEvent);
  },

  /**
   * Remove event listeners.
   * @private
   */
  _removeEventListeners() {
    this.dom.notificationLogTrigger.removeEventListener('click', this.listeners.clickTrigger);
    window.removeEventListener('SyspanelOpenEvent', this.listeners.SyspanelOpenEvent);
    window.removeEventListener('SyspanelTabEvent', this.listeners.SyspanelTabEvent);
    window.removeEventListener('SyspanelCloseEvent', this.listeners.SyspanelCloseEvent);

    this.hasActiveListeners.forEach((element) => {
      element.removeAllListeners();
    });
  },

  /**
   * Set dynamic notification data.
   * @public
   * @param {Object} notification - Notification data.
   * @param {string} notification.title
   * @param {string} [notification.description]
   * @param {Function} [notification.action]
   * @param {Date} [notification.timestamp]
   * @param {boolean} notification.isNew
   */
  setNotification(notification) {
    notification.isNew = true;
    this.state.isNew += 1;
    this.state.notifications.push(notification);
    this._showLiveNotification(notification);
    this._prependToLog(notification);
    this._updateCounter();
    this._removeEmptyMessage();

    /*
     * TODO: How to serve, place, handle audio files...
     * const audio = new Audio('public/vanilla-components/notification-log/notification.mp3');
     * audio.play();
     */
  },

  /**
   * Load notification log data.
   * @public
   * @param {Object[]} notifications - Notifications array.
   * @param {number} [notifications[].id]
   * @param {string} notifications[].title
   * @param {string} [notifications[].description]
   * @param {Function} [notifications[].action]
   * @param {Date} [notifications[].timestamp]
   * @param {boolean} notifications[].isNew
   */
  loadNotificationLog(notifications) {
    notifications.forEach((notification) => {
      if (notification.isNew) {
        this.state.isNew += 1;
        this._showLiveNotification(notification);
      }
      this.state.notifications.push(notification);
      this._prependToLog(notification);
    });

    this._updateCounter();
    this._removeEmptyMessage();
  },

  /**
   * Get notification log data.
   * https://gitlab.seld.rnd.ericsson.se/eds/eds/issues/307
   * @public
   * @return {Object[]} - Notifications array.
   */
  getNotifications() {
    return this.state.notifications;
  },

  /**
   * Set time after which notification will be removed.
   * @public
   * @param {number} ms - Time in milliseconds.
   */
  setNotificationTimeout(ms) {
    this.notificationTimeout = ms;
  },

  /**
   * Show dynamic notification.
   * @private
   * @param {Object} notification - Notification data.
   * @param {string} notification.title
   * @param {string} [notification.description]
   * @param {Function} [notification.action]
   * @param {Date} [notification.timestamp]
   * @param {boolean} notification.isNew
   */
  _showLiveNotification(notification) {
    this._removeLiveNotification();

    const newNotification = new Notification({
      title: notification.title,
      description: notification.description,
      action: notification.action,
      timeout: this.notificationTimeout,
    });
    newNotification.init();

    this.state.liveNotification = newNotification;

    newNotification.getNotification().addEventListener('click', () => {
      this._clearNotification();
    });
    this.hasActiveListeners.push(newNotification.getNotification());
  },

  /**
   * Remove dynamic notification.
   * @private
   */
  _removeLiveNotification() {
    if (this.state.liveNotification) {
      this.state.liveNotification.destroy();
    }
  },

  /**
   * Add notification to the beginning of the log.
   * @private
   * @param {Object} notification - Notification data.
   * @param {number} [notification.id]
   * @param {string} notification.title
   * @param {string} [notification.description]
   * @param {Function} [notification.action]
   * @param {Date} [notification.timestamp]
   * @param {boolean} notification.isNew
   */
  _prependToLog(notification) {
    const notificationLogItem = document.createElement('div');
    notificationLogItem.classList.add('notification-log-item');

    if (notification.isNew) {
      notificationLogItem.classList.add('new');
    }

    const titleElement = document.createElement('div');
    titleElement.classList.add('title');
    titleElement.innerText = notification.title;
    notificationLogItem.appendChild(titleElement);

    const descriptionElement = document.createElement('div');
    descriptionElement.classList.add('description');
    descriptionElement.innerText = notification.description;
    notificationLogItem.appendChild(descriptionElement);

    const timestampElement = document.createElement('span');
    timestampElement.classList.add('notification-log-item-time');
    timestampElement.innerText = this._getRelativeTime(new Date(notification.timestamp));
    notificationLogItem.appendChild(timestampElement);

    notificationLogItem.dataset.timestamp = notification.timestamp;
    this.state.notificationsDOM.push(notificationLogItem);
    const notificationContainer = document.querySelector('.notification-log-container');

    // Store index to connect DOM element with notification Object
    notificationLogItem.dataset.index = notificationContainer.querySelectorAll('.notification-log-item').length;
    notificationContainer.insertBefore(notificationLogItem, notificationContainer.firstChild);

    if (notification.action) {

      notificationLogItem.addEventListener('click', () => {
        window.dispatchEvent(this.events.toggleSyspanel);
        setTimeout(() => {
          notification.action();
        }, 250);
      });
      this.hasActiveListeners.push(notificationLogItem);
    }

  },

  /**
   * Update notification counter element.
   * @private
   */
  _updateCounter() {
    window.dispatchEvent(this.events.NotificationLogCounterUpdate);

    this.dom.notificationLogNumber.innerHTML = this.state.isNew;
    const hasNotification = 'has-notification';
    const noNotification = 'no-notification';

    if (this.state.isNew === 0) {
      if (this.dom.notificationLogTrigger.classList.contains(hasNotification)) {
        this.dom.notificationLogTrigger.classList.remove(hasNotification);
        this.dom.notificationLogTrigger.classList.add(noNotification);
      }
    } else {
      this.seen = false;
      this.dom.notificationLogTrigger.classList.remove(noNotification);
      this.dom.notificationLogTrigger.classList.add(hasNotification);
    }
  },

  /**
   * Mark notification as seen.
   * @private
   * @param {HTMLElement} item - Notification DOM element
   */
  _markAsSeen(item) {
    item.classList.remove('new');

    // Update state of related notification object
    this.state.notifications[item.dataset.index].isNew = false;
  },

  /**
   * Mark all notifications as seen.
   * @private
   */
  _markAllAsSeen() {
    for (let i = 0; i < this.state.notificationsDOM.length; i += 1) {
      this._markAsSeen(this.state.notificationsDOM[i]);
    }
    this.state.notificationsDOM = [];
  },

  /**
   * Handle click on notification log icon.
   * @private
   */
  _clickTrigger() {
    window.dispatchEvent(this.events.toggleSyspanel);
    this.seen = true;
    this._updateTimes();
  },

  /**
   * Remove 'new' state from clicked notification and update counter.
   * @private
   */
  _clearNotification() {
    this.state.isNew -= 1;
    this._updateCounter();
    this._markAsSeen(this.state.notificationsDOM[this.state.notificationsDOM.length - 1]);
  },

  /**
   * Remove 'new' state from all notifications and update counter.
   * @private
   */
  _clearAllNotifications() {
    if (this.seen) {
      this.state.isNew = 0;
      this.seen = false;
      this._updateCounter();
      this._markAllAsSeen();
    }
  },

  /**
   * Hide message that is displayed when log is empty.
   * @private
   */
  _removeEmptyMessage() {
    this.dom.notificationLogEmpty.classList.add('hidden');
  },

  /**
   * Get time relative to current date for example '1 mo' (1 month ago).
   * @private
   * @param {Date} timestamp - Date of notification.
   * @return {string} - Time in relative format.
   */
  // TODO: Talk about i18n and l10n, since right now time units are hardcoded...
  _getRelativeTime(timestamp) {
    const msPerMinute = 60 * 1000;
    const msPerHour = msPerMinute * 60;
    const msPerDay = msPerHour * 24;
    const msPerMonth = msPerDay * 30;
    const msPerYear = msPerDay * 365;

    const elapsed = Date.now() - timestamp.getTime();

    if (elapsed < msPerMinute) {
      return '';
    } else if (elapsed < msPerHour) {
      return Math.round(elapsed / msPerMinute) + ' m';
    } else if (elapsed < msPerDay) {
      return Math.round(elapsed / msPerHour) + ' h';
    } else if (elapsed < msPerMonth) {
      return Math.round(elapsed / msPerDay) + ' d';
    } else if (elapsed < msPerYear) {
      return Math.round(elapsed / msPerMonth) + ' mo';
    } else {
      return Math.round(elapsed / msPerYear) + ' y';
    }
  },

  /**
   * Update notification elements to display time in relative format.
   * @private
   */
  _updateTimes() {
    for (let i = 0; i < this.state.notificationsDOM.length; i += 1) {
      const item = this.state.notificationsDOM[i];
      item.querySelector('.notification-log-item-time')
        .innerHTML = this._getRelativeTime(new Date(item.dataset.timestamp));
    }
  },
};
