// @eds/vanilla/charts/bar-charts/HorizontalBarChartStacked.js
/**
 * @file Module to create Horizontal Bar Chart Stacked component.
 * @requires d3
 * @requires Stage
 * @requires Axis
 * @requires Legend
 * @requires ColorScale
 */

import * as d3 from 'd3';
import { Stage } from '../common/Stage';
import { Axis } from '../common/Axis';
import { Tooltip } from '../common/Tooltip';
import { Legend } from '../common/Legend';
import { ColorScale } from '../common/ColorScale';

/**
 * Class representing a horizontal bar chart stacked.
 */
export class HorizontalBarChartStacked {

  /**
   * Setup chart properties.
   * @param {Object} props - The properties to initialize the stage.
   * @param {Object} props.data - The data necessary to draw the chart.
   * @param {HTMLElement} props.element - The DOM element to append the chart.
   * @param {number} [props.height] - The height of the chart.
   * @param {Object} [props.margin] - The top, right, bottom and left margins.
   * @param {number} [props.margin.top] - The top margin.
   * @param {number} [props.margin.right] - The right margin.
   * @param {number} [props.margin.bottom] - The bottom margin.
   * @param {number} [props.margin.left] - The left margin.
   * @param {string} [props.x] - The X axis properties.
   * @param {string} [props.x.unit] - The unit of the X axis.
   */

  constructor(props) {
    this.data = props.data;
    this.element = props.element;
    this.margin = {
      left: 16,
      right: 24,
      top: 16,
      bottom: 32,
      ...props.margin,
    };

    this.height = props.height || 250;
    this.thresholds = props.thresholds;
    this.x = { unit: '', ...props.x };

    this.axisPadding = 8;
    this.fontSize = 14;
    this.barWidth = 6;
    this.animationSpeed = 400;
    this.valueSeparator = 2;

    if (this.x.unit) {
      this.margin.bottom += this.fontSize + this.axisPadding;
    }
  }

  /**
   * Initialize the chart.
   * @public
   */

  init() {
    this.stage = new Stage({
      element: this.element,
      margin: this.margin,
      height: this.height
    });

    this.stage.init();
    this.stage.onResize(this.draw.bind(this));

    this.canvas = this.stage.getCanvas();

    this.colorScale = new ColorScale({
      length: this.data.series.length
    });

    this.colorScale.init();

    this.legend = new Legend({
      element: this.element,
      data: this.data,
      draw: this.draw.bind(this),
      highlight: this._highlight.bind(this),
      margin: this.margin
    });
    this.legend.init();

    this.tooltip = new Tooltip({
      element: this.element,
      data: this.data,
      unit: {
        value: this.x.unit
      },
      hoverClass: 'line-hover'
    });

    this.tooltip.init();

    this.draw();
  }

  /**
   * Destroy the chart and associated events.
   * @public
   */

  destroy() {
    this.stage.destroy();
  }

  /**
   * Draw the chart based on data.
   * @public
   * @param {Object} data - The data necessary to draw the chart.
   */

  draw(data) {
    this.data = data || this.data;
    this.data = this.legend.filter(this.data);
    this.width = this.stage.getCanvasWidth();

    const y0 = d3.scaleBand()
      .domain(this.data.common)
      .rangeRound([this.axisPadding, this.height])
      .paddingInner(1)
      .paddingOuter(0.4);

    const xMin = d3.min(this.data.series, d => d3.min(d.values)) >= 0 ? 0 :
      d3.min(this.data.series, d => d3.min(d.values));
    let xMax = 0;

    for (let i = 0; i < this.data.common.length; i += 1) {
      let sum = 0;

      for (let y = 0; y < this.data.series.length; y += 1) {
        sum += this.data.series[y].values[i];
      }
      if (sum > xMax) {
        xMax = sum;
      }
    }

    const x = d3.scaleLinear()
      .domain([xMin, xMax])
      .nice()
      .range([0, this.width]);

    const axis = new Axis({
      position: 'bottom',
      canvas: this.canvas,
      x: x,
      width: this.width,
      height: this.height,
      margin: this.margin,
      fontSize: this.fontSize,
      axisPadding: this.axisPadding,
      tickFormat: (d) => d3.format('.2s')(d).replace('.0', ''),
      unit: this.x.unit
    });

    axis.init();

    const hasLabelSpace =
      this.height / this.data.common.length > this.fontSize + this.axisPadding + this.barWidth * 2;

    const barGroups = this.canvas
      .classed('bar-chart', true)
      .selectAll('.bar-group')
      .data(this.data.common);

    const barGroupsEnter = barGroups
      .enter()
      .append('g')
      .classed('bar-group', true);

    barGroupsEnter
      .append('text')
      .classed('label', true)
      .classed('hidden', !hasLabelSpace)
      .attr('x', '2')
      .attr('y', -this.barWidth - 2)
      .style('pointer-events', 'none')
      .text(d => d);

    barGroups
      .selectAll('.line-hover')
      .attr('x1', -this.width)
      .attr('x2', this.width);

    barGroupsEnter
      .merge(barGroups)
      .transition()
      .duration(this.animationSpeed)
      .attr('transform', d => `translate(${isNaN(x(0)) ? 1 : x(0) + 1}, ${y0(d)})`);

    const bars = barGroupsEnter
      .merge(barGroups)
      .selectAll('.bar')
      .data((d, i) => {
        const prevValues = [];
        return this.data.series.map((s, k) => {
          const getPrevious = () => {
            if (k === 0) {
              return 0;
            } else {
              prevValues.push(this.data.series[k - 1].values[i]);
              return prevValues.reduce((a, b) => a + b);
            }
          };
          return {
            key: s.name,
            value: s.values[i],
            previous: getPrevious()
          };
        });
      });

    bars
      .enter()
      .append('line')
      .merge(bars)
      .style('pointer-events', 'none')
      .attr('stroke-width', this.barWidth)
      .attr('class', d => 'bar ' + this.colorScale.get()(d.key))
      .transition()
      .duration(this.animationSpeed)
      .attr('x1', d => -(x(0) - x(d.previous)))
      .attr('x2', d => isNaN(d.value) ? 0 : -(x(0) - x(d.value) + x(0) - x(d.previous) + this.valueSeparator));

    barGroups
      .exit()
      .remove();

    bars
      .exit()
      .transition()
      .duration(this.animationSpeed)
      .attr('x1', 0)
      .attr('x2', 0)
      .remove();

    if (this.data.series.length > 0) {
      barGroupsEnter
        .append('line')
        .classed('line-hover', true)
        .attr('stroke-width', y0(this.data.common[1]) - y0(this.data.common[0]))
        .attr('stroke', 'blue')
        .attr('opacity', '0')
        .attr('x1', -this.width)
        .attr('x2', this.width)
        .on('mouseover', (group, i, selection) => {
          const all = barGroupsEnter.selectAll('.bar');
          const allLabels = barGroupsEnter.selectAll('.label');

          this.tooltip.update(i, this.data);

          all.attr('opacity', '.2');
          allLabels.attr('opacity', '.2');

          d3.select(selection[i].parentNode).selectAll('.bar').attr('opacity', '1');
          d3.select(selection[i].parentNode).selectAll('.label').attr('opacity', '1');
        })
        .on('mouseout', () => {
          const all = barGroupsEnter.selectAll('.bar');
          const allLabels = barGroupsEnter.selectAll('.label');

          all.attr('opacity', '1');
          allLabels.attr('opacity', '1');
        });
    }
  }

  /**
   * Highlight one common element in all series.
   * @private
   * @param {String} key - The key matching the common element.
   */

  _highlight(key) {
    const bars = this.canvas
      .selectAll('.bar');

    if (key) {
      bars.attr('opacity', d => d.key === key ? '1' : '.2');
    } else {
      bars.attr('opacity', '1');
    }
  }
}
